sdk/storage/storage-internal-avro/src/AvroParser.ts

```diff
@@ -0,0 +1,365 @@
+    return view.getFloat32(0, true); // littleEndian = true
+  }
+
+  public static async readDouble(stream: AvroReadable): Promise<number> {
+    const u8arr = await AvroParser.readFixedBytes(stream, 8);
+    const view = new DataView(u8arr.buffer, u8arr.byteOffset, u8arr.byteLength);
+    return view.getFloat64(0, true); // littleEndian = true
+  }
+
+  public static async readBytes(stream: AvroReadable): Promise<Uint8Array> {
+    const size = await AvroParser.readLong(stream);
+    if (size < 0) {
+      throw new Error("Bytes size was negative.");
+    }
+
+    return await stream.read(size);
+  }
+
+  public static async readString(stream: AvroReadable): Promise<string> {
+    const u8arr = await AvroParser.readBytes(stream);
+
+    // FIXME: need TextDecoder polyfill for IE
+    let utf8decoder = new TextDecoder();
+    return utf8decoder.decode(u8arr);
+  }
+
+  private static async readMapPair<T>(
+    stream: AvroReadable,
+    readItemMethod: (s: AvroReadable) => Promise<T>
+  ): Promise<KeyValuePair<T>> {
+    const key = await AvroParser.readString(stream);
+    // FIXME: what about readFixed which need a length as parameter.
+    const value = await readItemMethod(stream);
+    return { key, value };
+  }
+
+  public static async readMap<T>(
+    stream: AvroReadable,
+    readItemMethod: (s: AvroReadable) => Promise<T>
+  ): Promise<Record<string, T>> {
+    const readPairMethod = async (stream: AvroReadable): Promise<KeyValuePair<T>> => {
+      return await AvroParser.readMapPair(stream, readItemMethod);
+    };
+
+    const pairs: KeyValuePair<T>[] = await AvroParser.readArray(stream, readPairMethod);
+    let dict: Record<string, T> = {};
+    for (const pair of pairs) {
+      dict[pair.key] = pair.value;
+    }
+    return dict;
+  }
+
+  private static async readArray<T>(
+    stream: AvroReadable,
+    readItemMethod: (s: AvroReadable) => Promise<T>
+  ): Promise<T[]> {
+    let items: T[] = [];
+    for (
+      let count = await AvroParser.readLong(stream);
+      count != 0;
+      count = await AvroParser.readLong(stream)
+    ) {
+      if (count < 0) {
+        // Ignore block sizes
+        await AvroParser.readLong(stream);
+        count = -count;
+      }
+
+      while (count--) {
+        const item: T = await readItemMethod(stream);
+        items.push(item);
+      }
+    }
+    return items;
+  }
+}
+
+interface RecordField {
+  name: string;
+  type: string | ObjectSchema | (string | ObjectSchema)[]; // Unions may not immediately contain other unions.
+}
+
+interface ObjectSchema {
+  type: "record" | "enum" | "array" | "map" | "fixed";
+  name?: string;
+  aliases?: string;
+  fields?: RecordField[];
+  symbols?: string[];
+  values?: string;
+  size?: number;
+}
+
+export abstract class AvroType {
+  /**
+   * Reads an object from the stream.
+   *
+   * @param stream
+   */
+  public abstract read(stream: AvroReadable): Promise<Object | null>;
```

Comment:
Storage SDK doesn't return null. You can return undefined when no Object is returned.

It's not a perfect design to have undefined and null in JS language. It's easy to have issues when we both use undefined & null in our SDK. For consistency, please use undefined here.

Reply:
`null` value is part of the Avro spec.
https://avro.apache.org/docs/1.8.2/spec.html#schema_primitive

Reply:
Ok.